src/libraries/System.Text.Json/tests/Serialization/NumberHandlingTests.cs

```diff
@@ -0,0 +1,708 @@
+        }
+
+        private static void PerformEncodingTest<T>(List<T> numbers, JsonSerializerOptions options)
+        {
+            // All input characters are escaped
+            IEnumerable<string> numbersAsStrings = numbers.Select(num => GetNumberAsString(num));
+            string input = JsonSerializer.Serialize(numbersAsStrings, options);
+
+            // Unescaping works
+            List<T> deserialized = JsonSerializer.Deserialize<List<T>>(input, options);
+            Assert.Equal(numbers.Count, deserialized.Count);
+            for (int i = 0; i < numbers.Count; i++)
+            {
+                Assert.Equal(numbers[i], deserialized[i]);
+            }
+
+            // Every number is written as a string, and custom escaping is not honored.
+            string serialized = JsonSerializer.Serialize(deserialized, options);
+            var reader = new Utf8JsonReader(Encoding.UTF8.GetBytes(serialized));
+            reader.Read();
+            while (reader.Read())
+            {
+                if (reader.TokenType == JsonTokenType.EndArray)
+                {
+                    break;
+                }
+                else
+                {
+                    Assert.Equal(JsonTokenType.String, reader.TokenType);
+#if BUILDING_INBOX_LIBRARY
+                    Assert.False(reader.ValueSpan.Contains((byte)'\\'));
+#else
+                    foreach (byte val in reader.ValueSpan)
+                    {
+                        if (val == (byte)'\\')
+                        {
+                            Assert.True(false, "Unexpected escape token.");
+                        }
+                    }
+#endif
+                }
+            }
+        }
+
+        [Fact]
+        public static void NullableNumber_RoundtripNull()
+        {
+            PerformNullabilityTest<byte>();
+            PerformNullabilityTest<sbyte>();
+            PerformNullabilityTest<short>();
+            PerformNullabilityTest<int>();
+            PerformNullabilityTest<long>();
+            PerformNullabilityTest<ushort>();
+            PerformNullabilityTest<uint>();
+            PerformNullabilityTest<ulong>();
+            PerformNullabilityTest<float>();
+            PerformNullabilityTest<decimal>();
+            PerformNullabilityTest<byte?>();
+            PerformNullabilityTest<sbyte?>();
+            PerformNullabilityTest<short?>();
+            PerformNullabilityTest<int?>();
+            PerformNullabilityTest<long?>();
+            PerformNullabilityTest<ushort?>();
+            PerformNullabilityTest<uint?>();
+            PerformNullabilityTest<ulong?>();
+            PerformNullabilityTest<float?>();
+            PerformNullabilityTest<decimal?>();
+        }
+
+        private static void PerformNullabilityTest<T>()
+        {
+            string nullAsJson = "null";
+            string nullAsQuotedJson = $@"""{nullAsJson}""";
+
+            if (Nullable.GetUnderlyingType(typeof(T)) != null)
```

Comment:
Splitting this up into two tests (Nullable and non-Nullable) would avoid this `if` statement and make it more obvious what is being tested.